### ISSUE TITLE
CrossRef plugin

### ISSUE THREAD

**Issue by jasonpriem** [figshare/Total-Impact#21]
This should be the first plugin we bring up to the new spec, since a lot of other ones depend on it. It has to return these metrics:
- URLs (from DOIs)
- DOIs (from DOIs. I know this sounds wacky, but we need 'doi" as a response metric...the update controller will expect to find them there, not in artifact IDs, because we may get DOIs from reverse lookups, as well as from user input)

It would be nice to have it also return:
- whatever article metadata they've got (I don't know what it is...title, authors, date?)
- DOIs (from URLs -- referse lookup. This would allow folks to input URLs for articles, and we'd still have the DOIs for other plugins)
- DOIs and URLs from Pubmed IDs (some folks may have their articles listed this way) Looks like it's [fairly easy](http://biostar.stackexchange.com/questions/6137/how-can-i-find-a-corresponding-doi-for-a-given-pmid) to do for most pmids

Heather, I'm "assigning" it to you because you said you might have a lil time, but of course let me/us know if you want help.

**Comment by jasonpriem** [comment]
Hm, github isn't letting me assign this issue to @hpiwowar right now...so just remember it's yours :)

**Comment by hpiwowar** [comment]
sounds good.  I've assigned myself to it :)

**Comment by hpiwowar** [comment]
The crossref plugin is close to ready.  A few points:
- PMID->DOI and URL is in
- I don't know how to do URL->DOI.  Let's say this is off the table for alpha.  We could maybe do lookup for other lookups in the future (http://www.crossref.org/guestquery/) but let's not for now.
- What will input to the crossref API look like?  Will the "doi" and "url" and "pmid" dictionary entries will be FALSE and the Crossref plugin just takes its input from the <id> fields?  Then it returns metric_names called "doi" and "url" and "pmid" and "title" and "journal" etc, right?

**Comment by jasonpriem** [comment]
- Awesome, well done.
- I agree that reverse lookup is a nice-to-have only.
- Yes, the values for "doi", "url", and "pmid" will probably be FALSE. But the CrossRef plugin doesn't have to care whether they are or not, or whether there might be other key:value pair included under each artifact ID. it only pays attention to the IDs given, because that's all it need to do its job.
  - Yes, it returns an object like the one described under "Plugins return" on the [wiki](https://github.com/mhahnel/Total-Impact/wiki/Plugin-requirements), adding key:value pairs for "doi", "title", and whatever other goodies you get under each article.  So you'd add something like `return_obj.articles.<my id>.title = "I'm doin' Science: We measured some stuff and here's what we got."`
